File: src/mffs/item/mode/ItemModeCube.java
package mffs.item.mode;

import java.util.HashSet;
import java.util.Set;

import mffs.api.IProjector;
import mffs.render.model.ModelCube;
import net.minecraft.tileentity.TileEntity;

import org.lwjgl.opengl.GL11;

import universalelectricity.core.vector.Vector3;
import universalelectricity.prefab.vector.Region3;
import calclavia.lib.CalculationHelper;
import cpw.mods.fml.relauncher.Side;
import cpw.mods.fml.relauncher.SideOnly;

public class ItemModeCube extends ItemMode
{
	public ItemModeCube(int i, String name)
	{
		super(i, name);
	}

	public ItemModeCube(int i)
	{
		this(i, "modeCube");
	}

	@Override
	public Set<Vector3> getExteriorPoints(IProjector projector)
	{
		Set<Vector3> fieldBlocks = new HashSet<Vector3>();
		Vector3 posScale = projector.getPositiveScale();
		Vector3 negScale = projector.getNegativeScale();

		for (float x = -negScale.intX(); x <= posScale.intX(); x += 0.5f)
		{
			for (float z = -negScale.intZ(); z <= posScale.intZ(); z += 0.5f)
			{
				for (float y = -negScale.intY(); y <= posScale.intY(); y += 0.5f)
				{
					if (y == -negScale.intY() || y == posScale.intY() || x == -negScale.intX() || x == posScale.intX() || z == -negScale.intZ() || z == posScale.intZ())
					{
						fieldBlocks.add(new Vector3(x, y, z));
					}
				}
			}
		}
		return fieldBlocks;
	}

	@Override
	public Set<Vector3> getInteriorPoints(IProjector projector)
	{
		Set<Vector3> fieldBlocks = new HashSet<Vector3>();
		Vector3 posScale = projector.getPositiveScale();
		Vector3 negScale = projector.getNegativeScale();

		for (int x = -negScale.intX(); x <= posScale.intX(); x++)
		{
			for (int z = -negScale.intZ(); z <= posScale.intZ(); z++)
			{
				for (int y = -negScale.intY(); y <= posScale.intY(); y++)
				{
					fieldBlocks.add(new Vector3(x, y, z));
				}
			}
		}

		return fieldBlocks;
	}

	@Override
	public boolean isInField(IProjector projector, Vector3 position)
	{
		Vector3 projectorPos = new Vector3((TileEntity) projector);
		projectorPos.add(projector.getTranslation());
		Vector3 relativePosition = position.clone().subtract(projectorPos);
		CalculationHelper.rotateXZByAngle(relativePosition, -projector.getRotationYaw());
		CalculationHelper.rotateYByAngle(relativePosition, -projector.getRotationPitch());
		Region3 region = new Region3(projector.getNegativeScale().clone().multiply(-1), projector.getPositiveScale());
		return region.isIn(relativePosition);
	}

	@SideOnly(Side.CLIENT)
	@Override
	public void render(IProjector projector, double x, double y, double z, float f, long ticks)
	{
		GL11.glScalef(0.5f, 0.5f, 0.5f);
		ModelCube.INSTNACE.render();
	}

}

File: src/main/scala/mffs/field/mode/ItemModeCube.scala
package mffs.field.mode

import java.util.{HashSet, Set}
import javax.vecmath.Vector3d

import com.resonant.wrapper.lib.schematic.{Structure, StructureCube}
import mffs.api.machine.Projector
import nova.core.render.model.{BlockModelUtil, Model}
import nova.core.util.transform.{Cuboid, MatrixStack}

class ItemModeCube extends ItemMode {
	private val step = 1

	def getExteriorPoints(projector: IFieldMatrix): Set[Vector3d] = {
		val fieldBlocks = new HashSet[Vector3d]
		val posScale: Vector3d = projector.getPositiveScale
		val negScale: Vector3d = projector.getNegativeScale

		for (x <- -negScale.xi to posScale.xi by step; y <- -negScale.yi to posScale.yi by step; z <- -negScale.zi to posScale.zi by step)
			if (y == -negScale.yi || y == posScale.yi || x == -negScale.xi || x == posScale.xi || z == -negScale.zi || z == posScale.zi) {
				fieldBlocks.add(new Vector3d(x, y, z))
			}


		return fieldBlocks
	}

	def getInteriorPoints(projector: IFieldMatrix): Set[Vector3d] = {
		val fieldBlocks = new HashSet[Vector3d]
		val posScale = projector.getPositiveScale
		val negScale = projector.getNegativeScale

		//TODO: Check parallel possibility
		for (x <- -negScale.xi to posScale.xi by step; y <- -negScale.yi to posScale.yi by step; z <- -negScale.zi to posScale.zi by step)
			fieldBlocks.add(new Vector3d(x, y, z))

		return fieldBlocks
	}

	override def isInField(projector: IFieldMatrix, position: Vector3d): Boolean = {
		val projectorPos: Vector3d = new Vector3d(projector.asInstanceOf[TileEntity])
		projectorPos.add(projector.getTranslation)
		val relativePosition = position.clone.subtract(projectorPos)
		relativePosition.transform(new EulerAngle(-projector.getRotationYaw, -projector.getRotationPitch, 0))
		val region = new Cuboid(-projector.getNegativeScale, projector.getPositiveScale)
		return region.intersects(relativePosition)
	}

	override def getID: String = "modeCube"

	override def getStructure: Structure = new StructureCube

	override def render(projector: Projector, model: Model) {
		model.matrix = new MatrixStack().loadMatrix(model.matrix).scale(0.5, 0.5, 0.5).getMatrix
		BlockModelUtil.drawCube(model)
	}
}